warehouse/models/intermediate/payments/int_payments__settlements_summarized.sql

```diff
@@ -0,0 +1,80 @@
+{{ config(materialized = 'table',) }}
+
+WITH settlements AS (
+    SELECT *
+    FROM {{ ref('stg_littlepay__settlements') }}
+),
+
+-- TODO: decide whether refunds should be joined in as part of the creation of this table?
+-- mapping refunds to settlements may require using settlement_id, which is not available downstream of this model
+
+-- some rows have null settlement type (debit vs. credit)
+-- so we try to impute based on settlement order
+impute_settlement_type AS (
+    SELECT
+        * EXCEPT(settlement_type),
+        COALESCE(settlement_type,
+            -- when a settlement (aggregation_id / RRN) appears multiple times, it is generally because the subsequent lines are refunds
+            -- vast majority of refunds have exactly one debit line and one credit (refund) line, but there are a few oddities with multiple credit (refund) lines
+            CASE
+                WHEN ROW_NUMBER() OVER(PARTITION BY aggregation_id, retrieval_reference_number ORDER BY settlement_requested_date_time_utc) > 1 THEN "CREDIT"
```

Comment:
A couple questions about this logic:

1. Just how frequent is the condition in which we have to impute a `settlement_type` instead of being provided one? Does that frequency vary by instance/agency?
2. Are we fairly confident about the assumptions we're making here (that everything after the first line is a credit)?

I'm hopeful that the answers to both of these are encouraging, but I'd have some hesitation if it's a common case _and_ we aren't really sure whether interpreting later lines as credits is reliable, since the scale of `transaction_amount` values counted in the wrong direction could stack up quickly.

Reply:
Re#1, It looks like it's a recency thing -- newer agencies have settlement type and older ones don't (another reason I think we really need to figure out how to update schema). 

See:
```
SELECT
  settlement_type IS NOT NULL AS has_type,
  ARRAY_AGG(DISTINCT participant_id) AS agencies,
  COUNT(*) AS ct,
  MIN(settlement_requested_date_time_utc) AS first_dt,
  MAX(settlement_requested_date_time_utc) AS last_dt
FROM `cal-itp-data-infra.staging.stg_littlepay__settlements`
GROUP BY 1
```

Re #2, confirmed that every single row with an imputed type of CREDIT has a refund associated:
```
WITH s AS (SELECT
        * EXCEPT(settlement_type),
        COALESCE(settlement_type,
            CASE
                WHEN ROW_NUMBER() OVER(PARTITION BY aggregation_id, retrieval_reference_number ORDER BY settlement_requested_date_time_utc) > 1 THEN "CREDIT"
                ELSE "DEBIT"
            END
        ) AS settlement_type,
        settlement_type IS NULL AS imputed_type,
FROM `cal-itp-data-infra.staging.stg_littlepay__settlements`),

r AS (
  SELECT *
  FROM `cal-itp-data-infra.staging.stg_littlepay__refunds`
)

SELECT
  s.settlement_type,
  s.participant_id,
  s.imputed_type,
  s.settlement_id,
  s.aggregation_id,
  s.retrieval_reference_number,
  s.transaction_amount,
  r.refund_id,
  r.aggregation_id,
  r.retrieval_reference_number,
  r.proposed_amount,
  r.refund_amount
FROM s
FULL OUTER JOIN r
USING (retrieval_reference_number)
WHERE s.imputed_type and s.settlement_type = "CREDIT" AND r.refund_id IS NULL
```

Among data where we do have settlement type, there are no cases where there is more than one row with the same type for the same aggregation ID / RRN pair:

```
SELECT 
  aggregation_id,
  retrieval_reference_number,
  settlement_type,
  COUNT(*) AS ct
FROM `cal-itp-data-infra.staging.stg_littlepay__settlements`
WHERE settlement_type IS NOT NULL
GROUP BY 1,2,3
HAVING ct > 1
```

There are exactly three cases where there are more than 2 rows for an aggregation_id/RRN pair (so one type has to have a duplicate) where settlement type is missing and for those I cross referenced against the Elavon data and confirmed that the ultimate settled amounts corresponded to first row = debit and rest = credit (refund). 

There are exactly three cases where there is an aggregation ID/RRN that only appears once (i.e, only one row) that also appears in the refunds table **but all three actually have an explicit settlement_type of `DEBIT`**.

```
WITH s AS (SELECT
        * EXCEPT(settlement_type),
        COALESCE(settlement_type,
            CASE
                WHEN ROW_NUMBER() OVER(PARTITION BY aggregation_id, retrieval_reference_number ORDER BY settlement_requested_date_time_utc) > 1 THEN "CREDIT"
                ELSE "DEBIT"
            END
        ) AS settlement_type,
        settlement_type IS NULL AS imputed_type,
FROM `cal-itp-data-infra.staging.stg_littlepay__settlements`),

row_num AS (
  SELECT aggregation_id, retrieval_reference_number, COUNT(*) AS ct
  FROM s
  GROUP BY 1,2
),

r AS (
  SELECT *
  FROM `cal-itp-data-infra.staging.stg_littlepay__refunds`
)

SELECT
  s.settlement_type,
  s.participant_id,
  s.imputed_type,
  s.settlement_id,
  s.aggregation_id,
  s.retrieval_reference_number,
  s.transaction_amount,
  r.refund_id,
  r.aggregation_id,
  r.retrieval_reference_number,
  r.proposed_amount,
  r.refund_amount
FROM s
LEFT JOIN row_num USING(aggregation_id, retrieval_reference_number)
FULL OUTER JOIN r
USING (settlement_id)
WHERE row_num.ct = 1 AND refund_id IS NOT NULL
```

Reply:
It seems pretty solid to me, and I am not sure what the alternatives would be; I still intend to try to reconcile this with refunds & Elavon separately downstream, I just don't really think there's an alternative unless you see another way to impute -- the main other ways to do it would be cross referencing with refunds/Elavon but I think that's the next step after this table.

Reply:
That all sounds great; I just wanted to understand the grounding for the logic! Thanks for walking through your assessment.